### DISCUSSION TITLE
SC + Opensearch using Aggregation Spout

### DISCUSSION THREAD

**Discussion by dhaneshsabane** [DigitalPebble/storm-crawler#1091] [Q&A] [answered]
Hello all,

I've been working on setting up SC with Opensearch as the backend by referring the stormcrawler-docker repository. The setup was interesting and I've managed to get all the services up and running without any errors. However, my document in the `status` index on Opensearch which is my **spout** is not being picked up for crawling. I'm unable to figure out what I'm missing. Can someone help me figure out the issue please?

Here's the document I've added to the **status** index.

Request:
```
GET status/_search
{
    "query": {
        "match_all": {}
    }
}
```
Response:
```
{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 10,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "status",
        "_id": "-2wAYYoBDIBwcCXkH9jP",
        "_score": 1,
        "_source": {
          "url": "https://en.wikipedia.org/wiki/StormCrawler",
          "key": "crawl",
          "nextFetchDate": "2023-09-03T15:30:00Z"
        }
      }
    ]
  }
}
```

Worker logs only shows the following:
```
2023-09-04 16:50:44.034 c.d.s.o.p.AggregationSpout I/O dispatcher 65 [INFO] [spout #0]  ES query returned 0 hits from 0 buckets in 2 msec with 0 already being processed. Took Infinity msec per doc on average.
2023-09-04 16:50:44.060 c.d.s.o.p.AggregationSpout Thread-26-spout-executor[11, 11] [INFO] [spout #1]  Populating buffer with nextFetchDate <= 2023-09-04T16:50:44Z
2023-09-04 16:50:44.063 c.d.s.o.p.AggregationSpout I/O dispatcher 65 [INFO] [spout #1]  ES query returned 0 hits from 0 buckets in 2 msec with 0 already being processed. Took Infinity msec per doc on average.
```

**Comment by jnioche** [comment]
Hi @dhaneshsabane 
Judging by the log, you are using an older version of SC. The latest one (released yesterday) has quite a few improvements for OpenSearch, you should use it instead.

Looking at your problem, the key is an usual one. Did you inject the seed manually? You could change the log level and see what queries are generated by the spouts and compare with the one you posted above to check that there isn't anything missing from the document that is required by the spouts.
Maybe change your config to use a single spout instance, it will make it easier to find things in the logs.
The only thing I can think of would be a difference  between the number of shards defined for the index and the number of spout instances - if the latter is > 1 then they get assigned a specific shard, which means that if your index has more shards than you have instances of spouts, some will not get picked,

**Comment by dhaneshsabane** [comment]
Hey @jnioche !

> Hi @dhaneshsabane Judging by the log, you are using an older version of SC. The latest one (released yesterday) has quite a few improvements for OpenSearch, you should use it instead.
> 

Sounds good! I'll try the newer version.

> Looking at your problem, the key is an usual one. Did you inject the seed manually? 

Yes. I've injected that seed manually. My first insert on the index was only with the **url** but after seeing that it was not working, I tried to add the other fields in an attempt to make it work.

> You could change the log level and see what queries are generated by the spouts and compare with the one you posted above to check that there isn't anything missing from the document that is required by the spouts. Maybe change your config to use a single spout instance, it will make it easier to find things in the logs. The only thing I can think of would be a difference between the number of shards defined for the index and the number of spout instances - if the latter is > 1 then they get assigned a specific shard, which means that if your index has more shards than you have instances of spouts, some will not get picked,

I have two shards and the same number of spouts. I remember reading in some other discussions / answers / documentation that it is important to keep the number of spouts = number of index shards.

I'll enable the debug logs while also updating the version to see if it fixes the issue for me. If not, I'll come back with some follow up questions. Thanks for your help!

**Reply by dhaneshsabane** [reply]
@jnioche 

In my previous comment, I mentioned that I had 2 shards but that was wrong. I had 10 shards and only 2 spouts. So, clearly, there was a discrepancy. So I re-created the index with 2 shards as you can see from the following:

**Request**
```
GET _cat/shards/status/?v
```

**Response**
```
index  shard prirep state   docs store ip          node
status 1     p      STARTED    0  208b 172.25.0.14 opensearch-node1
status 1     r      STARTED    0  208b 172.25.0.15 opensearch-node2
status 0     r      STARTED    1 3.9kb 172.25.0.14 opensearch-node1
status 0     p      STARTED    1 3.9kb 172.25.0.15 opensearch-node2
```

After this, I once again manually inserted the seed -

```
POST status/_doc
{
  "url": "https://en.wikipedia.org/wiki/StormCrawler",
  "nextFetchDate": "2023-09-03T15:30:00Z"
}
```

However, the spout is still unable to fetch the URL. I turned on DEBUG log to find the query that the spout is running.

```
2023-09-05 15:26:36.171 c.d.s.o.p.AggregationSpout Thread-26-spout-executor[11, 11] [DEBUG] [spout #1]  OpenSearch query SearchRequest{searchType=QUERY_THEN_FETCH, indices=[status], indicesOptions=IndicesOptions[ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, expand_wildcards_hidden=false, allow_aliases_to_multiple_indices=true, forbid_closed_indices=true, ignore_aliases=false, ignore_throttled=true], routing='null', preference='_shards:1|_local', requestCache=null, scroll=null, maxConcurrentShardRequests=0, batchedReduceSize=512, preFilterShardSize=null, allowPartialSearchResults=null, localClusterAlias=null, getOrCreateAbsoluteStartMillis=-1, ccsMinimizeRoundtrips=true, source={"from":0,"size":0,"query":{"bool":{"filter":[{"range":{"nextFetchDate":{"from":null,"to":"2023-09-05T15:26:36Z","include_lower":true,"include_upper":true,"boost":1.0}}}],"adjust_pure_negative":true,"boost":1.0}},"explain":false,"track_total_hits":-1,"aggregations":{"partition":{"terms":{"field":"key","size":50,"min_doc_count":1,"shard_min_doc_count":0,"show_term_doc_count_error":false,"order":[{"top_hit":"asc"},{"_key":"asc"}]},"aggregations":{"docs":{"top_hits":{"from":0,"size":2,"version":false,"seq_no_primary_term":false,"explain":false,"sort":[{"nextFetchDate":{"order":"asc"}},{"url":{"order":"asc"}}]}},"top_hit":{"min":{"field":"nextFetchDate"}}}}}}, cancelAfterTimeInterval=null, pipeline=null}
```

Translating this to an actual index query using the Dev Tools on the dashboards gives me the following:

Request:
```
GET status/_search
{
  "query": {
    "bool": {
      "filter":[
        {
          "range": {
            "nextFetchDate": {
              "from":null,
              "to":"2023-09-05T15:26:36Z",
              "include_lower":true,
              "include_upper":true,
              "boost":1.0
            }
          }
        }
      ]
    }
  }
}
```

Response:

```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 2,
    "successful": 2,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 0,
    "hits": [
      {
        "_index": "status",
        "_id": "NXDwZYoBDIBwcCXk_cXA",
        "_score": 0,
        "_source": {
          "url": "https://en.wikipedia.org/wiki/StormCrawler",
          "nextFetchDate": "2023-09-03T15:30:00Z"
        }
      }
    ]
  }
}
```

I'm running out of ideas, here. I'm really struggling to identify what is missing in this whole equation. Any help is appreciated. 🙏🏽

**Comment by jnioche** [comment, marked as answer]
The actual query is 

```json
{
  "from": 0,
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "range": {
            "nextFetchDate": {
              "from": null,
              "to": "2023-09-05T15:26:36Z",
              "include_lower": true,
              "include_upper": true,
              "boost": 1
            }
          }
        }
      ],
      "adjust_pure_negative": true,
      "boost": 1
    }
  },
  "explain": false,
  "track_total_hits": -1,
  "aggregations": {
    "partition": {
      "terms": {
        "field": "key",
        "size": 50,
        "min_doc_count": 1,
        "shard_min_doc_count": 0,
        "show_term_doc_count_error": false,
        "order": [
          {
            "top_hit": "asc"
          },
          {
            "_key": "asc"
          }
        ]
      },
      "aggregations": {
        "docs": {
          "top_hits": {
            "from": 0,
            "size": 2,
            "version": false,
            "seq_no_primary_term": false,
            "explain": false,
            "sort": [
              {
                "nextFetchDate": {
                  "order": "asc"
                }
              },
              {
                "url": {
                  "order": "asc"
                }
              }
            ]
          }
        },
        "top_hit": {
          "min": {
            "field": "nextFetchDate"
          }
        }
      }
    }
  }
}
```

the doc needs a _key_ field but the doc you added is missing one. Note that URLs belonging to the same domain or host need to be routed using the key so that they end up in the same shard.

I'd recommend that you inject with a local topology (see latest version of resources generated by the archetype) and not manually.

**Reply by dhaneshsabane** [reply]
That's it! I managed to fix one but messed up the other! 😅 

Thank you so much for your guidance @jnioche !

Can you point me to any documentation around the *key* field which can help me understand how it is used for routing?

**Reply by jnioche** [reply]
See https://github.com/DigitalPebble/storm-crawler/blob/f7dfa823ef1292abfa29424e509b187301970b56/external/opensearch/src/main/java/com/digitalpebble/stormcrawler/opensearch/persistence/StatusUpdaterBolt.java#L227

it uses a URLPartitioner internally which has the following default configuration

https://github.com/DigitalPebble/storm-crawler/blob/f7dfa823ef1292abfa29424e509b187301970b56/core/src/main/resources/crawler-default.yaml#L44C1-L45C31

**Reply by dhaneshsabane** [reply]
Thanks once again! 👍🏽
